Reference verifier section(s) revamp (#576)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

~~~ pseudocode
FUNC mm_to_em(mm: measurement-map) -> element-map {
    em := element-map::NEW()

    IF mm.mkey:
        em.element-id = mm.mkey

    em.element-claims = mm.mval

    RETURN em
}

FUNC mms_to_ems(mms: [ + measurement-map ]) -> [ + element-map ] {
    ems := [ + element-map ]::NEW()

    FOREACH mm IN mms:
        em := mm_to_em(mm)
        ems::APPEND(em)

    RETURN ems
}
~~~
{: #algo-mm-to-em title="Transform Measurement Map(s) into Element Map(s)"}

**Claim Names.**
The combination of `environment`, optional `element-id`, and map key within each `element-claims` encodes the name of an Element Claim.
The value of the corresponding map element represents an atom of actual state.
This specification does not assign special meanings to any Claim name, it only specifies rules for determining whether two Claim names are the same.

**Merge Rules.**
If two Element ECTs have the same `environment`, `cmtype`, `authority` and `profile` then their `element-list`s are merged.
Any duplicates MUST be pruned.

Two `measurement-values-map` containing duplicate codepoints and with non-equivalent measurement values MUST NOT be merged.

> These are effectively two different acceptable states that need to be processed separately.

Reviewer: Would `element-map` be more accurate than `measurement-values-map` in this context?

Author: Yes, definitely.  Thanks for spotting it.

Author: This has been addressed in 3fdfcf7.  However, it would probably be worth calling out `element-id` as one of the matching criteria.

Reviewer 2: Addressed

If all the attributes which are present in the C-ECT `environment` (e.g., `instance-id` or `group-id`) are also present in the ACS-ECT and are binary identical, the two environments match.
Otherwise, the environments do not match.

Any attribute that is present in the ACS-ECT but not in the C-ECT is ignored in the comparison.

##### Authority Comparison {#sec-compare-authority}

The comparison between a C-ECT's `authority` value and an ACS-ECT's `authority` value is as follows: if every entry in the C-ECT `authority` matches byte-by-byte an entry in the ACS-ECT `authority`, the authorities match.
Otherwise, the authorities do not match.

The order of the items within each `authority` does not affect the result of the comparison.

When comparing two `$crypto-key-type-choice` items for equality, the processor MUST treat them as equal if their deterministic CBOR encoding is binary equal.

##### Element List Comparison {#sec-compare-element-list}

A C-ECT's `element-list` matches an ACS-ETC's `element-list` if all the `element-map`s in the C-ECT's `element-list` match the `element-map`s in the ACS-ECT's `element-list`.

> Any `element-map` that is present in the ACS-ECT's `element-list` but not in the C-ECT's is ignored in the comparison.

Reviewer: I've been looking at CoRIM modeling with others and modeling multiple accepter attester states is a sticking point (i.e.  modeling ref values such that SPDM index 1 can have digest value A or B). Can you spot my mistake in the following logic?

After ECT merging is complete, there will be exactly one ECT with a given environment, cmtype, authority, and profile. The element-list will effectively contain a set of all unique element-maps. Borrowing some of your notation from another MR, this is means that the below forms are equivalent:
```
T = [
  E=e_0,
  T=reference-value, # added the cmtype as "T"
  [
    M{ K=k_0, V=v_0 },
    M{ K=k_0, V=v_1 }
  ]
]

T_1 = [
  E=e_0,
  T=reference-value,
  [
    M{ K=k_0, V=v_0 }
  ]
]
+
T_2 = [
  E=e_0,  # Was e in original comment. Based on the context, I believe e_0 was intended.
  T=reference=value,
  [
    M{ K=k_0, V=v_1 }
  ]
]
```
So, due to merging, a CoRIM author can provide either T or T1 + T2. From now on, we'll work with the merged form T.

From here, it would seem that the final form T doesn't support the case where the author intended to represent multiple acceptable states. That is:
```
# This can be matched:
T = [
  E=e_0,
  T=evidence,
  [
    M{ K=k_0, V=v_0 },
    M{ K=k_0, V=v_1 }
  ]
]

# This is should be matched, but seemingly would not be:
T = [
  E=e_0,
  T=evidence,
  [
    M{ K=k_0, V=v_0 }
  ]
]
```
This is due to the statement that **all** element-maps in the C-ECT must match the element-maps in the ACS-ECT. `M{ K=k_0, V=v_1 }` would not match because it is (rightfully) missing from the evidence.

I may be missing a key concept here. For example, could it be the case that T1 and T2 were actually not merged because one of their element-maps were not the same? That could be a solution to my dilemma, but I'm not sure what other issues that might cause.

Author: I believe we have a bug in the spec.

The idea should be that alternative states must be encoded in separate triples, and that merging rules must disallow coalescing triples that represent alternative states.

Otherwise, as you have also noticed, the matching algorithm does not work because the `rv-item.condition` becomes self-contradictory.

I think that the sentence in §5.1.5:

> Each ref-claims entry can represent a different possible state of the Environment

needs to go.

Reviewer 2: Now captured in this issue: https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/575

If, after visiting all the entries in the ACS, the matched entries set is not empty, the C-ECT matches the ACS.
Conversely, if the matched entries set is empty, the C-ECT does not match the ACS.

If the C-ECT contains a profile and the profile defines an algorithm for a given codepoint, the processor MUST use the algorithm defined by the profile in comparisons involving that codepoint.
If the condition ECT contains a profile, but the profile does not define an algorithm for a particular codepoint, the processor MUST use the standard algorithm described in this document for comparisons involving that codepoint.

The specific comparisons performed depend on the type of relation being processed.
In general, the processor will perform comparisons based on `environment` (see {{sec-compare-environment}}) as well as more specialized comparisons based on the type of ECT matched.
Element ECTs will match on `element-list` (see {{sec-compare-element-list}}), Key ECTs will match on `key-list` (<cref>[TODO]</cref>), and Domain ECTs will typically match on `children` (see {{sec-compare-environment}}).

> <cref>
> [TBC]
> Are authority checks always carried out?
> </cref>

Author: ```suggestion
<cref>
[TBC]
Check, for each relation processing, where authority comparisons apply.
</cref>
```

The algorithm used to compare two values of a `measurement-values-map` attribute depends on the attribute's type.

The processor needs to select the appropriate algorithm for the given attribute, including any extensions defined by a supported profile.

Non-negative codepoints represent standard data representations.
The comparison algorithms for these are defined in this document (in the sections below) or in other specifications.
Some attributes allow different types.
These are designated by a CBOR tag that allows clear type disambiguation.

Negative codepoints represent profile-defined data representations.
The processor MUST use the attribute name, the profile associated with the condition ECT, and, if present, the CBOR tag value to select the comparison algorithm.

If the processor is unable to determine the applicable comparison algorithm for an attribute, it MUST behave as though the C-ECT does not match the ACS-ECT.

Profile writers SHOULD use CBOR tags for widely applicable comparison methods to ease Verifier implementation compliance across profiles.
<cref>
[TBC]
Unclear what this recommendation aims to achieve.

> </cref>

Author: ```suggestion
```

In a Domain ECT, it is the parent environment to which the child domains are related.
In a Key ECT, it is the environment to which the keys belong.
In all cases, Environments are identified using instance, class, or group identifiers.

* `authority`: Identifies the entity that issued the tuple.
The authority of a given ECT is typically established through a digital signature on the Claim.
For instance, a signature of the authoritative supply chain entity over the CoRIM containing the triple from which the ECT was obtained, or the Attesting Environment that signed the Evidence from which the ECT is derived.
It is represented as the key material by which the authority (and corresponding provenance) of the tuple can be determined.
A typical example is the authority's PKIX certificate.
This is a mandatory attribute in an ECT.

* `profile`: The profile that defines the domain of interpretation of this tuple.
This is the `profile` attribute of the CoRIM that contained the original triple from which this ECT was obtained.
This is an optional attribute in an ECT.
If no profile is used, the attribute is omitted.
<cref>
[TBC]
What profile applies in such a case?

> </cref>

Author: ```suggestion
If no profile is used, the attribute is omitted, and the only comparison rules that apply are those specified in {{sec-comparison-rules}}.
```

    em.element-claims = mm.mval

    RETURN em
}

FUNC mms_to_ems(mms: [ + measurement-map ]) -> [ + element-map ] {
    ems := [ + element-map ]::NEW()

    FOREACH mm IN mms:
        em := mm_to_em(mm)
        ems::APPEND(em)

    RETURN ems
}
~~~
{: #algo-mm-to-em title="Transform Measurement Map(s) into Element Map(s)"}

**Claim Names.**
The combination of `environment`, optional `element-id`, and map key within each `element-claims` encodes the name of an Element Claim.
The value of the corresponding map element represents an atom of actual state.
This specification does not assign special meanings to any Claim name, it only specifies rules for determining whether two Claim names are the same.

**Merge Rules.**
If two Element ECTs have the same `environment`, `cmtype`, `authority` and `profile` then their `element-list`s are merged.
Any duplicates MUST be pruned.

<cref>
[TBC]
Original text also states: "If two merged `measurement-values-map` contains duplicate codepoints and the measurement values are not equivalent, then the Verifier SHALL report an error and stop validation processing."
Are we sure this is the case?
Can't they be interpreted as alternative states?

> </cref>

Author: ```suggestion
Two `measurement-values-map` containing duplicate codepoints and with non-equivalent measurement values MUST NOT be merged.
These are effectively two different acceptable states that need to be processed separately.
```

1. The addition ECT is added to the ACS.

Note that this new ACS item is essentially a copy of the matched evidence ECT, which has been re-asserted under the authority of the Reference Value Provider.

#### Endorsed Values Augmentation (Phase 4)

Endorsed values augmentation is the process of adding Claims about the Attester's actual state with Endorser authority rather than Attester authority.
Augmentation is predicated on a certain condition matching the actual state currently encoded in the ACS.

##### Processing `ev` Relations

Endorsed Values and Conditional Endorsed Values are matched with ACS entries by iterating through the `ev` list.
For each `ev` entry, the condition ECT is compared with an ACS ECT with `cmtype` 0, 1 or 2 (i.e., reference-values, endorsements or evidence).
If the two match, the addition ECT is added to the ACS.

<cref>
[TBC]
Original text also states: "Some condition values can match against multiple ACS-ECTs, or sets of ACS-ECTs. If there are multiple matches, then each match is processed independently from the others."  Why should add exact copies of the same addition ECT?  Is there any specific semantics associated with clones?

> </cref>

Author: ```suggestion
```

However, ECTs of different types can sometimes be compared if the comparison is based on common attributes ({{fig-ect-common}}).
See {{sec-proc-dm}} for an example of this, where the `environment`s of `D-ECT`s and `E-ECT`s are compared.

Conceptually, the processor creates a "matched entries" set and populates it with all ACS entries that match the C-ECT.
If, after visiting all the entries in the ACS, the matched entries set is not empty, the C-ECT matches the ACS.
Conversely, if the matched entries set is empty, the C-ECT does not match the ACS.

If the C-ECT contains a profile and the profile defines an algorithm for a given codepoint, the processor MUST use the algorithm defined by the profile in comparisons involving that codepoint.
If the condition ECT contains a profile, but the profile does not define an algorithm for a particular codepoint, the processor MUST use the standard algorithm described in this document for comparisons involving that codepoint.

The specific comparisons performed depend on the type of relation being processed.

> In general, the processor will perform comparisons based on `environment` (see {{sec-compare-environment}} and `authority` (see {{sec-compare-authority}}), as well as more specialized comparisons based on the type of ECT matched.

Author: ```suggestion
In general, the processor will perform comparisons based on `environment` (see {{sec-compare-environment}}) as well as more specialized comparisons based on the type of ECT matched.
```

If they match, the `evs` series array is iterated.
For each series entry, if the selection ECT matches an ACS ECT, the addition ECT is added to the ACS.
Series iteration terminates either when the first matching series entry has been processed, or when no series entries match.

##### Processing `keys` Relations

Keys relations identify cryptographic keys that require additional key verification steps.

Keys are matched with ACS entries by iterating through the `keys` list.
For each `keys` entry, the condition ECT is compared with an ACS ECT with `cmtype` 1 or 2 (i.e., endorsements or evidence).
If they match, perform the following steps for each key in the condition ECT `key-list`:

1. Verify the certificate signatures for each certificate in the certification path.
1. Verify the revocation status for each certificate in the certification path.
1. Verify the key usage restrictions that are appropriate for `key-type`.

A key that successfully passes the above checks is said to be verified.

Add all the verified keys to the addition ECT's `key-list` and add the addition ECT to the ACS.

> ##### Processing `dm` Relations {#sec-proc-dm}

Reviewer 3: Processing `dm` relations, is a double loop, where if `any` Domain Member ECT is added into the ACS, then The Outer Loop is re-iterated.

This core concept is missing in the new PR, which needs addition.

Author: The comment can be resolved as we clarified this during the weekly call.  Yogesh: Feel free to unresolve if you disagree.

Domain Membership relations describe the expected topological arrangement of the Attester.

Domains are matched with ACS entries by iterating through the `dm` list.

The following algorithm assumes that the graph described by the condition ECTs in the `dm` relation is acyclic.

> It also assumes that the `dm-item`s in the `dm` relation are topologically sorted.

Author: ```suggestion
It also assumes that the `dm-item`s in the `dm` relation are topologically sorted (bottom up, from leaves to root).
```

This allows the algorithm to execute in one pass.

For each `dm` entry, the condition ECT is compared with either an ACS Element ECT with `cmtype` 0 (i.e., corroborated evidence) or a Domain ECT with `kind` 0 (i.e., member).
<cref>
[TBC]
The original text says: "Domain Membership ECTs (i.e., cmtype equals domain-member) in the dm staging area are matched with ACS entries where cmtype is set to evidence, reference-values i.e. corroborated evidence, or domain-member [...]" but it seems that matching evidence ECTs is superfluous and potentially leading to the wrong result (i.e., matching uncorroborated environments).

> </cref>

Author: ```suggestion
For each `dm` entry, the condition ECT is compared with either an ACS Element ECT with `cmtype` 2 (i.e., evidence) or a Domain ECT with `kind` 0 (i.e., member).
```

## The CoRIM Processor {#sec-corim-processor}

This document assumes that Verifier implementations will differ.
In order to describe normative Verifier behaviour, this section presents a reference Verifier and illustrates how the data is utilized within the appraisal phases detailed in {{sec-appraisal-phases}}.
If the Verifier operates on CoRIM documents, it is RECOMMENDED that it follows this algorithm.

### High-Level View

The RATS Verifier takes Evidence, Reference Values, Endorsements and an Appraisal Policy for Evidence as inputs, and produces Attestation Results as output.
{{fig-verifier-internal}} illustrates how the CoRIM processor fits into the wider RATS Verifier architecture.

~~~ aasvg
{::include pics/corim-verifier.txt}
~~~
{: #fig-verifier-internal artwork-align="center" title="CoRIM Processing Flow"}

The CoRIM processor accepts Reference Values and Endorsements in the form of CoRIM documents, as well as Evidence that has been converted into a CoRIM-compatible format using transforms such as those described in {{-rats-evidence-trans}}.

> Before the appraisal can begin, all Conceptual Messages must be broken down and reshaped into a common internal representation.

Reviewer 3: ```suggestion
Before the appraisal can begin, all  CoRIMs must be broken down and reshaped into a common internal representation.
```

> The internal representations of Reference Values and Endorsements are stored in a staging area prior to appraisal initiation.

Reviewer 3: How about Attest Key Triples and Device Identity Triples & Domain Membership and Domain Dependency Triples.

> Instead, the internal representation of Evidence is used to initialize the ACS.

Reviewer 3: ```suggestion
The Appraisal process is initiated upon processing Evidence and transforming it into an internal representation. The ACS is then initialized  using the internal representation of Evidence.
```

Each Attester has a different ACS.
The Verifier ensures that Evidence inputs are associated with the correct ACS.
All the internal representations are based on the Environment-Claim Tuple (ECT).
The ECT is the core data structure used to represent both Claims and matching conditions during appraisal by the CoRIM processor.
The CoRIM processor algorithm loads items from the staging area one by one and applies the required condition-matching rules against the ECTs in the ACS.
If the match is successful, the ACS is "augmented" with Claims from the matched item.
Once all the items in the staging area have been processed, the state of the Attester, as understood by the CoRIM processor, is reflected in the ACS.
The computed ACS can then be handed over to subsequent appraisal phases, such as Appraisal Policy evaluation and Attestation Results computation, repackaging and signing.

### Data Structures

This section describes the data structures used by the CoRIM processor.

#### ECT {#sec-ect}

The Environment-Claim Tuple is a core internal construct of the CoRIM Verifier.
It is used to describe a feature (or "Claim") of the appraised environment alongside relevant metadata.

> All ECTs, except those containing Evidence Claims, are typically obtained from CoMID triples.

Reviewer 3: Is this true, how about the CoTLs ??

If the `condition` containing the "reference" ECTs matches the Evidence ECTs, the Evidence ECTs are re-asserted with the RVP authority carried in the `addition` ECT and the `cmtype` set to `reference-values`.
The re-asserted ECTs are added to the ACS.
Refer to {{sec-proc-rv}} for how the `rv` entries are processed.

{{fig-rv-ect-cond}} shows the profiled Element ECT for a Reference Values condition.

~~~ cddl
{::include cddl/intrep-ect-refval-condition.cddl}
~~~
{: #fig-rv-ect-cond title="Profiled ECT for Reference Values (condition)"}

{{fig-rv-ect-add}} shows the profiled Element ECT for a Reference Values addition.

~~~ cddl
{::include cddl/intrep-ect-refval-addition.cddl}
~~~
{: #fig-rv-ect-add title="Profiled ECT for Reference Values (addition)"}

As this is used to corroborate an Evidence ECT, its layout is identical to that of an `Evidence-addition-ECT`.
The only differences are the values of the `element-list`, `authority` and `cmtype` attributes.
Here, the `authority` value is that of the RVP rather than the Attester, and the `cmtype` value is `reference-values` rather than `evidence`.

> Furthermore, the `element-list` is created if a match is found during processing, rather than when the triple is transformed into the internal representation.

Reviewer 3: I would re-phrase it slightly better,

Furthermore the `element-list` in the addition-ECT entry is empty when the triple is transformed into its internal representation. Only during appraisal processing, if the conditions are satisfied then the element-list is created. See section (XXX Corroboration of Reference Values)

Therefore, it may be absent.

##### Endorsed Values {#sec-ir-endval}

The internal representation of Endorsed Values uses the `ev` and `evs` relations.
These are lists of ECTs that describe matching conditions and the additions that apply when these conditions are met.

The `ev` relation ({{fig-ev}}) applies to Endorsed Values (EV) and Conditional Endorsement (CE) triples.

~~~ cddl
{::include cddl/intrep-ev.cddl}
~~~
{: #fig-ev title="Endorsed Values Relation"}

The `ev` relation compares the condition ECTs with those in the ACS.
If all the ECTs are found in the ACS, the addition ECTs are added to it.
Note that when the `ev` relation is for an EV triple, the optional `element-list` inside the condition is not used; however, it is used for CE triples.

The `evs` relation ({{fig-evs}}) applies to Conditional Endorsement Series (CES) Triples.

~~~ cddl
{::include cddl/intrep-evs.cddl}
~~~
{: #fig-evs title="Endorsed Values Series Relation"}

The `evs` relation compares the condition ECTs with the ACS.
<cref>

> [TBC]

Reviewer 3: Clarify Line 2165 during the review process

##### Keys {#sec-ir-keys}

The internal representation of Attest Key and Device Identity triples uses the `keys` relation ({{fig-k}}), whereby each `key-item` corresponds to either an `attest-key-triple-record` or an `identity-triple-record`.

~~~ cddl
{::include cddl/intrep-key.cddl}
~~~
{: #fig-k title="Keys Relation"}

<cref>
[TODO]
Specialise condition/addition ECTs.
Define constraints.
</cref>

##### Domains {#sec-ir-domains}

The internal representation of domains uses a common `domain-item` structure:

~~~ cddl
{::include cddl/intrep-domain-item.cddl}
~~~
{: #fig-dm-item title="Domain Item"}

The internal representation of Domain Membership uses the `dm` relation ({{fig-dm}}), whereby each `domain-item` corresponds to a `domain-membership-triple-record`.

~~~ cddl
{::include cddl/intrep-domain-mem.cddl}
~~~
{: #fig-dm title="Domain Membership Relation"}

The internal representation of Domain Dependency uses the `dd` relation ({{fig-dd}}), whereby each `domain-item` corresponds to a `domain-dependency-triple-record`.

~~~ cddl
{::include cddl/intrep-domain-dep.cddl}
~~~
{: #fig-dd title="Domain Dependency Relation"}

> {{fig-domain-ect-cond}} shows the profiled Domain ECT for both Domain Membership and Dependency conditions.

Reviewer 3: We should be careful in call Domain Condition as `ECT`, as strictly speaking it is NOT an ECT as it does not contain an Environment - but ONLY children a.ka. members.

If they match, perform the following steps for each key in the condition ECT `key-list`:

1. Verify the certificate signatures for each certificate in the certification path.
1. Verify the revocation status for each certificate in the certification path.
1. Verify the key usage restrictions that are appropriate for `key-type`.

A key that successfully passes the above checks is said to be verified.

Add all the verified keys to the addition ECT's `key-list` and add the addition ECT to the ACS.

##### Processing `dm` Relations {#sec-proc-dm}

Domain Membership relations describe the expected topological arrangement of the Attester.

Domains are matched with ACS entries by iterating through the `dm` list.

The following algorithm assumes that the graph described by the condition ECTs in the `dm` relation is acyclic.
It also assumes that the `dm-item`s in the `dm` relation are topologically sorted (bottom up, from leaves to root).
This allows the algorithm to execute in one pass.

For each `dm` entry, the condition ECT is compared with either an ACS Element ECT with `cmtype` 2 (i.e., evidence) or a Domain ECT with `kind` 0 (i.e., member).
All other ECTs are ignored.

> If all the `children` environments in the condition ECT have a matching ECT in the ACS, the ECT addition is added to the ACS.

Reviewer 3: ```suggestion
If all the `children` environments in the condition ECT of the `dm entry` match a corresponding `environment` in an ACS ECT, then the addition ECT of the `dm entry` is added to the ACS.
```

Domains are matched with ACS entries by iterating through the `dm` list.

The following algorithm assumes that the graph described by the condition ECTs in the `dm` relation is acyclic.

> It also assumes that the `dm-item`s in the `dm` relation are topologically sorted (bottom up, from leaves to root).

Reviewer 3: Please also add a reference to a clean topological sorting algorithm which can be referenced here. So far our internal algorithms were solely based on the logic documented in this draft. Now we rely on foreign algorithms to do correct job, before this draft algorithms can work properly.

Reviewer 2: afaik, there is no need to pursue this change any further. @yogeshbdeshpande, could you resolve it?
